### ISSUE TITLE
Reconsider the strong `Redeemer` / `Verifier` separation

### ISSUE THREAD

**Issue by JoshOrndorff** [Off-Narrative-Labs/Tuxedo#27]
In the current design, Tuxedo has a very strong separation between the logic that allows a single input to be consumed and the logic that checks whether a transaction is valid. This issue is to consider the pros and cons of that separation and decide to what extent it should be kept.

## Note on Terminology

Before going on, I'll observe that I _accidentally_ used the same terms as [IOHK's abstract model](https://eprint.iacr.org/2018/469.pdf) with different meanings :facepalm:

| Tuxedo Term | Abstract Model Term | Definition |
| ----------- | ------------------- | ---------- |
| Verifier    | None - token logic hard coded | The logic used to determine whether the consumed inputs and created outputs meet all necessary constraints |
| Redeemer    | Verifier            | The logic used to determine whether an individual input can be consumed |
| Witness     | Redeemer            | The data, likely a signature, that satisfies the logic to consume an input

### Pro 1: Writing pieces is very simple

With the current separation in place, the author of a piece doesn't need to care anything about whether individual inputs can be consumed. This is entirely handled by tuxedo executive. By the time the execution reaches the individual piece all that is left is typed data for the piece to check.

### Con 1: Pieces can't forcefully consume inputs

In some cases it is useful to consume a UTXO for housekeeping purposes.

Example 1: Consider a coupon with an expiration date. This UTXO would likely be guarded by a `CheckSig` so only the owner can use it. But if the expiration date passes, anyone should be able to clean up the state.

Example 2: In the proof of existence piece, two users could claim the same hash. When these competing claims are discovered anyone should be able to clean up the latter claim because the former is more legitimate. This is encountered in the [current codebase](https://github.com/Off-Narrative-Labs/Tuxedo/blob/7f1dabe37ce8621deb9df47d102ea775bbe2f763/frameless-runtime/src/poe.rs#L149-L180)

This can be addressed with a notion of evictions. See below.

### Con 2: Dynamic ownership is not immediately intuitive

In FRAME, pallet sudo is commonly used to guard functionality behind a signature by the `sudo` key. This is similar to hoe `SigCheck` works except that the key is dynamic. In Tuxedo this cannot be implemented as a redeemer because the sudo key is stored in state (a utxo) but redeemers do not have access to other state.

This could also be solved with evictions. (Again, see below).

### Con 3: Paying a private user

In many cases, a Tuxedo piece may want to pay assets to a specific private user. Consider a a treasury that wants to make a payment to a user, or a NFT game that charges some fees and allocates them to the original developer. When the output redeemers are not available, this cannot be done.

### Evictions
The idea of evictions is to add a field to the transaction that is similar to inputs or peeks (see #16), but the evicted inputs will be consumed regardless of whether their redeemer is satisfied. They will be forcefully removed.

This solves **Con 1** by allowing anyone to call, for example, a `CleanupExpiredCoupon` verifier or a `ResolvePoeDispute` verifier which will forcefully remove the outdated state.

This also solves **Con 2** with a little creative design. For an output that is intended to only be spent by root, make the redeemer `NobodyCanEverSpend`, then write a `SpendByRoot`  that verifier evicts it as long as the payload to the verifier contains a signature by the sudo key.

### Pass full outputs to verifier

Another alternative is to change the interface so that entire Outputs, including the redeemer, are passed to the verifiers. This solves **Con 3** but has the disadvantage that the Verifier must now know which redeemers are available to the aggregate runtime.

### Pass full transaction to verifier

We could weaken the separation even more by passing the entire transaction to the verifier. This has the all of the disadvantages of passing just the full outputs. It also has the disadvantage that it is the verifiers responsibility to check the redeemers. It has the advantage that evictions are no longer necessary because any verifier can simply choose to ignore the redeemer.

### Both of the above

It may be possible, to expose all of these APIs to piece developers, letting them choose which to implement. Pieces that can get by with a simple API can use it, and ones that require more powerful but more complex and dangerous APIs can use them. This may be possible with blanket implementations. Something like.

```rust
trait SimpleVerifier {
  // --snip--
}

trait PowerfulVerifier {
  // --snip--
}

impl<T: SimpleVerifier> PowerfulVerifier for T {
  // --snip--
}
```

**Comment by coax1d** [comment]
I wouldnt mind taking this.

How about the following solution:
```rust
trait SimpleVerifier {
    type Error;
    
     fn verify(
        &self,
        input_data: &[DynamicallyTypedData],
        output_data: &[DynamicallyTypedData],
    ) -> Result<TransactionPriority, OuterVerifierError>;
}

trait Verifier {
    type Error;
    
    fn verify<R: Redeemer>(
        &self,
        inputs: &[Input],
        outputs: &[Output<R>],
   ) -> Result<TransactionPriority, OuterVerifierError>;
}

// Default Implementation for a Verifier Type which only implements Simple Verifiers
pub struct MyVerifier;
impl<MyVerifier: SimpleVerifier> for MyVerifier {
    type Error = VerifierError;
    
    fn verify<R: Redeemer>(
        &self,
        inputs: &[Input],
        outputs: &[Output<R>],
   ) -> Result<TransactionPriority, OuterVerifierError> {
       let input_data: Vec<DynamicallyTypedData> = inputs
            .iter()
            .map(|i| {
                TransparentUtxoSet::<R>::peek_utxo(&i.output_ref)
                    .expect("We just checked that all inputs were present.")
                    .payload
            })
            .collect();
          
        let output_data: Vec<DynamicallyTypedData> = outputs
            .iter()
            .map(|o| o.payload.clone())
            .collect();
        
        <MyVerifier as SimpleVerifier>::verify(self, input_data, output_data)
   }
    
}

```

What we could do is add a derive proc macro too every verifier so that if one really only wants to have a `SimpleVerifier` then we they are required to add `#[derive(Verifier)]` above their verifier type. This way they can write their simple verifier and not worry about this other abstraction which will be the same for every `VerifierType`.

This would mean that in the executive we change it to instead of doing the `DynamicallyTypedData` extraction from the inputs and outputs to just instead pass the inputs and outputs to the `Verifier` plus the Redeemer type. What is also nice is that we still allow the `Executive` to do its checks such as Signature verification of the redeemers in the inputs.

Let me know what you think

**Comment by JoshOrndorff** [comment]
As I thought about this the past few days, I've come to nearly the same conclusion. I think it is a good idea to have the simple and full variants of the trait, and I like how you sketched it out.

Regarding the naming convention, I think we should stop calling this the verifier though because the word verifier means something different in the IOHK research (That was my silly mistake). I suggest something like this which is the same as yours except for the names.

```
pub trait ConstraintChecker<R: Redeemer> {
    type Error;

    fn check(
        &self,
        inputs: &[Input],
        outputs: &[Output<R>],
     ) -> Result<TransactionPriority, Self::Error>;
}
```

I think we can actually do it without a macro though. I'll make a PR to your PR to sketch it and see if you like the idea.

**Comment by coax1d** [comment]
@JoshOrndorff Should we keep this open until we verify that #41 solves the issue? Actually I just realized I didnt rename it as you wished.

**Comment by coax1d** [comment]
I know that our terms dont match up to the IOHK paper but they make intuitive sense and I sort of like them. How do you feel about keeping them the same? I think the terms are well defined in our documentation and design. I am open to either or though.

**Comment by JoshOrndorff** [comment]
Fixed in #41 and #45
